src/esp/gfx/replay/Player.cpp

```diff
@@ -25,17 +62,22 @@ void Player::readKeyframesFromJsonDocument(const rapidjson::Document& d) {
 Keyframe Player::keyframeFromString(const std::string& keyframe) {
   Keyframe res;
   rapidjson::Document d;
-  d.Parse<0>(keyframe.c_str());
+  d.Parse<0>(keyframe.data(), keyframe.size());
   esp::io::readMember(d, "keyframe", res);
   return res;
 }
 
-Player::Player(const LoadAndCreateRenderAssetInstanceCallback&
-                   loadAndCreateRenderAssetInstanceCallback,
-               const ChangeLightSetupCallback& changeLightSetupCallback)
-    : loadAndCreateRenderAssetInstanceCallback(
-          loadAndCreateRenderAssetInstanceCallback),
-      changeLightSetupCallback(changeLightSetupCallback) {}
+Keyframe Player::keyframeFromStringUnwrapped(
+    const Cr::Containers::StringView keyframe) {
+  Keyframe res;
+  rapidjson::Document d;
+  d.Parse<0>(keyframe.data(), keyframe.size());
+  esp::io::fromJsonValue(d, res);
+  return res;
+}
+
+Player::Player(AbstractPlayerImplementation& implementation)
+    : implementation_{implementation} {}
```

Comment:
Why not? 
```suggestion
Player::Player(AbstractPlayerImplementation implementation)
    : implementation_{std::move(implementation)} {}
```

Reply:
Fair enough.

Reply:
wait am confused, won't this still create duplicate instance? The AbstractPlayer caller and `_implementation` or am I misunderstanding something?

Reply:
`implementation_` is a reference as well :)

Reply:
Actually I went with a shared_ptr in the end, see https://github.com/facebookresearch/habitat-sim/pull/1947/commits/e85af83fb02f843a4b9a7687db62ea1db22c861c for details.